MFLT-5058 Add documentation for OTA (#145)
memfault/memfault-docs: docs/platform/ota.mdx

- Memfault users have different ways to configure which payloads to distribute
  to which devices.
- Ultimately, this process is meant to replace or update the software running on
  a device.

Read more about integrating with Memfault's OTA Updates system on each supported
platform:

- [OTA Updates for MCU devices](/docs/mcu/releases-integration-guide),
- [OTA Updates for Android (AOSP)](https://docs.memfault.com/docs/android/android-releases-integration-guide),
- [OTA Updates for Embedded Linux](https://docs.memfault.com/docs/linux/linux-releases-integration-guide).

Memfault provides two Release entities to solve the challenges of over-the-air
updates: Full Releases and Delta Releases.

### Full Releases

Full Releases provide a way to upgrade device Software Versions regardless of
the version the device is currently on.

If we define the version of the most current Release as B and every version
lower than B as `*`, we can describe the behavior of a Full Release as: `[*→B]`.

See the [updating behavior](#updating-behavior) section for more details.

#### Must-Pass-Through Full Releases

Full Releases can be marked as _must-pass-through_, or MPT for short. When a MPT

> Full Release is activated, a device on a lower version will be forced to "pass

Author: ![image](https://user-images.githubusercontent.com/16181067/145567174-55c417ed-35d5-42b0-9e4f-0ba26e1c1de0.png)
Assuming this applies here too and I should leave it at `MPT Release`.

through" this Release before an update is allowed to a Release with a version
greater than that of the MPT Full Release. Typically this setting is not needed.

### Delta Releases

A Delta Release describes a transition from exactly one From Version to exactly

> one Target Version. The behavior can be described as: `[A→B]`.

Author: Rename across doc to "To Version"

> one Target Version. The behavior can be described as: `[A→B]`.

Reviewer: Change to "To Version"?

Reviewer: Also, let's avoid "A" and "B" but use concrete numbers.

A Delta Release is only applied when the device’s current version matches the
Release’s From Version.

Delta Releases allow for more fine-grained control over the update paths of
devices. Additionally, Delta Releases generally contain smaller payloads,
reducing required bandwidth and costs associated with rolling out updates.

> Note that Delta Releases are referred to as [incremental updates in Android
> documentation][incremental], and likely elsewhere, too.

Reviewer: I'd pull this paragraph up to become the second statement after "the behavior can be described as".

And then:
```suggestion
Delta Releases are referred to as [incremental updates in Android
documentation][incremental], or "differential updates" in other places.
```

See the [updating behavior](#updating-behavior) section for more details.

> ### Activating a Release

Reviewer: I'd further call out that it's best practice to combine OTA with metrics/chart comparison to observe the effect the new software has.

For risk management, you'd want to activate releases on smaller test cohorts first and maybe even combine it with a staged rollout while observing the development of key metrics.

Author: I actually added a "best practices" section after "activating a Release":

![image](https://user-images.githubusercontent.com/16181067/145573775-b04c3227-7cb0-4e56-b047-170654b17b4e.png)

> You can activate a Release for a Cohort of your choice. When activating a

Reviewer: I'd also mention that releases are shareable between cohorts so that you can test a deployment in a beta cohort first before activating the *same* release in the production cohort.

Release for a Cohort, you can choose between two behaviors:

#### Normal Activation of a Release

Releases activated _normally_ are available immediately to all devices in a
Cohort.

#### Staged Rollout of a Release

> Staged-rollout deployments can be used to make a Release available to a Cohort

Reviewer: To reduce risk…

Author: ```
Staged-rollout deployments can be used to make a Release available to a Cohort
incrementally, reducing the risk of introducing bugs to a greater portion of
the fleet by limiting the number of devices that will install a new Release.
```

incrementally. When a staged-rollout deployment is created for a Cohort, a
random subset of its devices (amounting to the selected rollout percentage) will
be marked as _staged_ in this Cohort. Staged-rollout deployments are only
applicable to devices that are _staged_ in the Cohort.

> ## Updating Behavior

Reviewer: This might also be a great place to call out our per-device introspection feature that show the outcome of all the rues applied.

Author: It's at the end of the doc. Maybe I could keep that and link to it from here.

When a device queries the [Latest Release endpoint][endpoint], the Memfault

> platform needs to decide which OTA Payload to send it. This sections covers the

Reviewer: "send it" – do we mention the built-in use of a world-wide CDN anywhere? Maybe worth calling out in the intro paragraph and again here.
Bottom line: our OTA system is great for nightly CI deployments in the test lab during development to millions of sealed devices in the field.

behavior of our backend when receiving one such query.

### Applicability of an OTA Release

Along with the query to the Latest Release endpoint, a device sends:

- the Device ID (device serial) to determine its Cohort,
- the [Software Type][terminology] to determine the set of Releases against
  which to apply these rules,
- its current [Software Version][terminology],
- device-specific attributes such as [Hardware Version][terminology].

> A Release must be [active](#activating-a-release) to be applicable.

Reviewer: …for the cohort a given device is in…

To determine if an active Release is applicable to a device, its payload
attributes are tested against the parameters sent by the device. If they don’t
match, a Release is considered to be incompatible and treated as if it did not
exist.

An Release activated as a [staged-rollout](#staged-rollout-of-a-release) will
only be applicable to the _staged_ portion of the Cohort it targets. In other

> words, if the device is not _staged_ in its Cohort, such Releases will be

Author: will not be

applicable.

### Choosing Among Multiple Applicable OTA Releases

If a device has multiple possible update paths, the following rules of
precedence apply to determine which release is selected.

The following rules apply to all active Releases (Releases that have an active
Deployment) per Cohort:

1. Use a Delta Release whose From Version matches the current version of the
   device.
2. Otherwise, use a must-pass-through (MPT) Full Release with the lowest Target
   Version that’s greater than the current version of the device.
3. Otherwise, use a Full Release with the highest Target Version greater than
   the current version of the device.

In short, the precedence can be summarized like so: **Delta Release** > **MPT
Full Release** > **Full Release**.

Note that it is possible to skip active MPT Releases by defining a Delta Release

> that targets a Target Version higher than the Target Version of the MPT Release.

Reviewer: Do we need this?

## Managing OTA Updates with Memfault

The Memfault application contains OTA-related features in several of its views.

### OTA Releases

The **OTA Releases** section of the Memfault application can be accessed from
the main menu on the left side. In it, you’ll be able to access a list view of
all Full Releases and the equivalent list view for Delta Releases by selecting
one of the tabs under the title of the page.

Let's take a look at how to create Releases using the Memfault application.

:::important

Please note that the Release version you choose is expected to match the
Software Version a device will report once it has installed the OTA payload for
this Release.

:::

#### Creating a Full Release

To create a Full Release, click on the _Create Release_ button on the top-right

> area of the **OTA Releases** view. Among other information, you'll be promped

Author: prompted
